### PR TITLE
Lint as framework automatically

### LINKED ISSUE #2912: [Frameworks] Extend linter for module_name
For frameworks support we should extend the linter a bit:

- [x] Lint that `module_name` is a valid C99 identifier
- [ ] If an explicit `module_name` is specified, build both a static library and a framework

see also the discussion in CocoaPods/Core#205

### PR DESCRIPTION
**Pull request by neonichu**
Close #2912

- [x] Spec
- [x] Changelog

### CODE REVIEW

**Comment by neonichu** [comment]
Not sure how to do a spec for this, any ideas welcome

File: lib/cocoapods/validator.rb [hunk 1]
```diff
@@ -303,6 +303,7 @@ def tear_down_validation_environment
     # for all available platforms with xcodebuild.
     #
     def install_pod
+      use_frameworks = @use_frameworks || spec.attributes_hash.has_key?('module_name')
```
Comment (kylef) [inline lib/cocoapods/validator.rb]: I'm not sure if using module_map as the key decision to determining if we should lint as a framework makes sense. Explicitly defining a `module_name` will be rare as using the pod name would be the normal option.

I think we should also determine if the pod is Swift, and in that case use frameworks as that's how it will be integrated.

**Comment by kylef** [comment]
What about linting as a framework regardless (providing the pod supports iOS > 8.0 or OS X)? All pods from now on should support frameworks.

**Comment by neonichu** [comment]
Swift Pods already are implicitly opted in to `use_frameworks!` and therefore linted as such.

**Comment by neonichu** [comment]
I don't think we can in good faith lint all Pods as frameworks, because that would block updates for anything which doesn't work as a framework.

**Comment by neonichu** [comment]
Maybe we could actually switch this around and make linting as framework the default, with an option to lint as a static library. Most pods should work as framework after all.

**Comment by neonichu** [comment]
Updated this to lint as framework by default and make linting as a static lib optional.

File: lib/cocoapods/command/lib.rb [hunk 2]
```diff
@@ -113,8 +113,8 @@ def self.options
            ['--subspec=NAME', 'Lint validates only the given subspec'],
            ['--no-subspecs', 'Lint skips validation of subspecs'],
            ['--no-clean', 'Lint leaves the build directory intact for inspection'],
-           ['--use-frameworks', 'Lint uses frameworks to install the spec'],
-           ['--sources=https://github.com/artsy/Specs,master', 'The sources from which to pull dependant pods ' \
```
Comment (segiddins) [inline lib/cocoapods/command/lib.rb]: undo this change

File: lib/cocoapods/command/spec/lint.rb [hunk 3]
```diff
@@ -20,8 +20,8 @@ def self.options
            ['--subspec=NAME', 'Lint validates only the given subspec'],
            ['--no-subspecs', 'Lint skips validation of subspecs'],
            ['--no-clean', 'Lint leaves the build directory intact for inspection'],
-           ['--use-frameworks', 'Lint uses frameworks to install the spec'],
-           ['--sources=https://github.com/artsy/Specs,master', 'The sources from which to pull dependant pods ' \
```
Comment (segiddins) [inline lib/cocoapods/command/spec/lint.rb]: undo this change

**Comment by segiddins** [comment]
Also requires a change for `trunk push` and `repo push`?

**Comment by neonichu** [comment]
@segiddins I guess - seemingly they haven't got the `--use-frameworks` option. @mrackwitz was that intentional for some reason?

**Comment by segiddins** [comment]
Please also update the validator attribute to reflect the fact that frameworks are the default as well?

**Comment by kylef** [comment]
Needs tests and CHANGELOG entry.

**Comment by mrackwitz** [comment]
@neonichu @segiddins: Regarding `[trunk|repo] push`, I don't see a reason to not expose the option there. Especially with linting everything as framework, we'll need to allow to opt-out there.

**Comment by neonichu** [comment]
@mrackwitz :+1: - I'll create a corresponding PR for that later
